php-frontend/src/test/java/org/sonar/php/symbols/MethodSymbolTest.java

```diff
@@ -56,22 +56,92 @@ public void method_in_anonymous_class() {
+        "public function method1($a) {}", // is implemented from I1 by extending A1
+        "public function method2($a) {}", // is implemented from I2 through I1 by extending A1
+        "public function method3($a) {}}", // is unknown because I4 is unknown
+      "class C2 extends A3 {public function method4($a) {}}", // is unknown because A2 is unknown
+      "class C3 implements I3 {public function method5($a) {}}", // is not implemented because not declared in I3
+      "class C4 extends A2 {public function method6($a) {}}"); // is not implemented because A2 does not implement an interface
+
+    assertThat(classes.get("C1").getDeclaredMethod("method1").isOverriding()).isEqualTo(Trilean.TRUE);
+    assertThat(classes.get("C1").getDeclaredMethod("method2").isOverriding()).isEqualTo(Trilean.TRUE);
+    assertThat(classes.get("C1").getDeclaredMethod("method3").isOverriding()).isEqualTo(Trilean.UNKNOWN);
+    assertThat(classes.get("C2").getDeclaredMethod("method4").isOverriding()).isEqualTo(Trilean.UNKNOWN);
+    assertThat(classes.get("C3").getDeclaredMethod("method5").isOverriding()).isEqualTo(Trilean.FALSE);
+    assertThat(classes.get("C4").getDeclaredMethod("method6").isOverriding()).isEqualTo(Trilean.FALSE);
+  }
+
+  @Test
+  public void catch_dead_loop_in_isDeclaredInInterface() {
+    Map<String, ClassSymbol> classes = parseMultipleClasses("<?php",
+      "interface I1 extends I2 {}",
+      "interface I2 extends I1 {}",
+      "class C1 implements I1 {public function method1(){}}",
+      "class C2 extends C3 {public function method2(){}}",
+      "class C3 extends C2 {}");
+
+    assertThat(classes.get("C1").getDeclaredMethod("method1").isOverriding()).isEqualTo(Trilean.FALSE);
+    assertThat(classes.get("C2").getDeclaredMethod("method2").isOverriding()).isEqualTo(Trilean.FALSE);
+  }
+
+  @Test
+  public void private_method_does_not_override() {
+    Map<String, ClassSymbol> classes = parseMultipleClasses("<?php",
+      "class C1 {public function method1(){}}",
+      "class C2 extends C1 {private function method1(){}}");
+
+    assertThat(classes.get("C2").getDeclaredMethod("method1").isOverriding()).isEqualTo(Trilean.FALSE);
+  }
+
+  @Test
+  public void private_method_can_not_be_overridden() {
+    Map<String, ClassSymbol> classes = parseMultipleClasses("<?php",
+      "class C1 {private function method1(){}}",
+      "class C2 extends C1 {public function method1(){}}");
+
+    assertThat(classes.get("C2").getDeclaredMethod("method1").isOverriding()).isEqualTo(Trilean.FALSE);
+  }
+
+  @Test
+  public void stored_overriding_result() {
+    Map<String, ClassSymbol> classes = parseMultipleClasses("<?php",
+      "class C1 {public function method1(){}}",
+      "class C2 extends C1 {public function method1(){}}");
+
+    MethodSymbol methodSymbol = classes.get("C2").getDeclaredMethod("method1");
+    assertThat(methodSymbol.isOverriding()).isEqualTo(methodSymbol.isOverriding());
+  }
```

Comment:
I don't think that this test really tests that we cache the result and it would be hard to test. I would remove this test.